tokio/tests/sync_once_cell.rs

```diff
@@ -0,0 +1,97 @@
+#![warn(rust_2018_idioms)]
+#![cfg(feature = "full")]
+
+use tokio::runtime::Runtime;
+use tokio::sync::{AlreadyInitializedError, Lazy, NotInitializedError, OnceCell};
+use tokio::time::sleep;
+
+use std::future::Future;
+use std::pin::Pin;
+use std::time::{Duration, Instant};
+
+fn func1() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+    async fn f() -> u32 {
+        5
+    }
+    Box::pin(f())
+}
+
+fn func2() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+    async fn f() -> u32 {
+        sleep(Duration::from_secs(1)).await;
+        10
+    }
+    Box::pin(f())
+}
+
+#[test]
+fn get_or_init() {
+    let rt = Runtime::new().unwrap();
+
+    static ONCE: OnceCell<u32> = OnceCell::new();
+
+    rt.block_on(async {
+        let result1 = rt
+            .spawn(async { ONCE.get_or_init(func1).await })
+            .await
+            .unwrap();
+
+        let result2 = rt
+            .spawn(async { ONCE.get_or_init(func2).await })
+            .await
+            .unwrap();
```

Comment:
I'm missing a test that does this, but where the second call uses a method that panics.